### ISSUE TITLE
FR: track task creation time

### ISSUE THREAD

**Issue by mjpieters** [Nukesor/pueue#356]
#### Is your feature request related to a problem?

I wanted to track how long my tasks are waiting in the queue, only to find out there is no 'task created' timestamp. Only the time that the task started running, the time the task stopped running, and, if enqueued with a delay, the enqueue_at time are tracked.

#### Describe the solution you'd like

Add a 'created' timestamp to the task structure. It only needs to exist in the JSON output as far as I am concerned.

#### Describe alternatives you've considered

I now record the time I added a job to pueue separately, to combine this later on. This is error prone and more complex to handle.

**Comment by Nukesor** [comment]
I don't see any problems with this :)

If we track the `created_at` time, I would even go ahead and track the `enqueued_at` time as well. 
The `enqueue_at` time is currently only saved in the task's state and not persistant. Once the task is automatically enqueued, this information is lost.

- `created_at: Datetime` should be fairly straight forward. Just set the field in the `Task` constructor.
- `enqueued_at: Option<Datetime>` however has a few places where it needs to be set.
    - When running `enqueue`
    - When adding a task without `--stashed`
    - When restarting a task
    - When stashing a task (reset it to `None`)
    - Probably somewhere I'm currently not thinking of

**Comment by dmgolembiowski** [comment]
@mjpieters + @Nukesor, I'll give this a shot tonight.

**Comment by Nukesor** [comment]
Nah. That test is rather important as it ensures backward compatibility of states to old versions.

The proper way around it would be to define default deserialization values for the missing fields.
Checkout [serde defaults](https://serde.rs/attr-default.html)

That way, old states might have incorrect values, but at least they can be parsed and used as usual.

**Comment by dmgolembiowski** [comment]
> Thanks for taking care of this issue :)
You're welcome; I see a ton of potential for pueue.

> The proper way around it would be to define default deserialization values for the missing fields. Check out [serde defaults](https://serde.rs/attr-default.html).
> 
> That way, old states might have incorrect values, but at least they can be parsed and used as expected.

Sweet, I'll go ahead and work that out later tonight

**Comment by Nukesor** [comment]
We can also make this optional columns, that can be included once the query/filter syntax lands :)

Could be an interesting thing for some users.

**Comment by Nukesor** [comment]
ping @dmgolembiowski 

Do you still plan to work on this :)? I intend to do a release in the upcoming weeks and this is something I would like to see in that release!

**Comment by dmgolembiowski** [comment]
> ping @dmgolembiowski 
> 
> Do you still plan to work on this :)? I intend to do a release in the upcoming weeks and this is something I would like to see in that release^^. In case you're no longer interested, I'll probably continue working on this :)

Whoops! Sorry I, I've been busy with other things and forgot about this.

**Comment by dmgolembiowski** [comment]
Apologies if we're briefly losing commits due to an error I recently encountered:

```sh
╭─david at david-laptop in /home/david
╰─λ git clone https://github.com/dmgolembiowski/pueue/
Cloning into 'pueue'...
remote: fatal: bad tree object e0f56fb8b0421f874b4edd5c227912d219b669d9
remote: aborting due to possible repository corruption on the remote side.
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```

**Comment by dmgolembiowski** [comment]
Pardon, @Nukesor , could you peek at [github.com/dmgolembiowski/pueue/](https://github.com/dmgolembiowski/pueue/) and let me know what kinds of tests you'd like to integrate. As it's currently implemented, all tests are passing on nightly. (woo)

**Comment by Nukesor** [comment]
@dmgolembiowski I added a few comments on your commits, since the repository referenced in the PR no longer exists.

Regarding tests, I would be ok with adding a few statements in the tests that check enqueing and adding tasks :)

I.e.
- `test_normal_add` for `created_at`.
- `test_enqueued_tasks` and `test_delayed_tasks` for `enqueue_at`


Since we cannot check for exact times, we'll have to make do with a appropriate time range (~500ms-1000ms)

**Comment by Nukesor** [comment]
One more thing we should probably test:

Restarting tests without the `--in-place` option. That will also create a new (cloned) task.

One other thing I'm not sure about:  
Should editing a task's command/path update the `created_at` field? @mjpieters I would like to get your opinion on this as well :).

**Comment by mjpieters** [comment]
> Should editing a task's command/path update the `created_at` field?

Nope. It’s the same task still as far as I am concerned. If I need it to have a new created time I’ll create a new task…

**Comment by Nukesor** [comment]
Ping @dmgolembiowski

How is it going :)? In case you are no longer working on this, I'll work on this instead.
This should definitely be in the next release. I'm going on holiday next week and I'll try to polish everything for a 3.0 release during that time.

**Comment by dmgolembiowski** [comment]
> Ping @dmgolembiowski
> 
> How is it going :)? In case you are no longer working on this, I'll continue to work on this :)
> This should definitely be in the next release. I'm going on holiday next week and I'll try to polish everything for a 3.0 release during that time.

Awesome, sounds good to me.

**Comment by Nukesor** [comment]
Implemented in #386
